File: src/main/java/com/EmosewaPixel/pixellib/multiblocks/MultiblockTE.java
package com.EmosewaPixel.pixellib.multiblocks;

import com.EmosewaPixel.pixellib.blocks.RotatableMachineBlock;
import com.EmosewaPixel.pixellib.recipes.AbstractRecipeList;
import com.EmosewaPixel.pixellib.recipes.SimpleMachineRecipe;
import com.EmosewaPixel.pixellib.tiles.AbstractRecipeBasedTE;
import javafx.util.Pair;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.tileentity.TileEntityType;
import net.minecraft.util.math.BlockPos;

import java.util.List;
import java.util.Map;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public abstract class MultiblockTE<T extends SimpleMachineRecipe> extends AbstractRecipeBasedTE<T> {
    protected List<IMultiblockPart> partList;

    public MultiblockTE(TileEntityType type, AbstractRecipeList<T, ?> recipeList) {
        super(type, recipeList);
    }

    abstract protected MultiblockPatternBuilder getPattern();

    protected boolean isValidStructure() {
        BlockState state = getBlockState();
        Map<BlockPos, Predicate<BlockState>> pattern = getRotatedPattern();
        final BlockPos posInPattern = pattern.entrySet().stream().filter(e -> e.getValue().test(state)).map(Map.Entry::getKey).findFirst().get();

        if (posInPattern == null)
            return false;

        partList = pattern.keySet().stream().filter(pos -> world.getTileEntity(pos) instanceof IMultiblockPart).map(pos -> (IMultiblockPart) world.getTileEntity(pos)).collect(Collectors.toList());

        return pattern.keySet().stream().allMatch(pos -> pattern.get(pos).test(world.getBlockState(getPos().subtract(posInPattern).add(pos))));
    }

    protected Predicate<Block> selfPredicate() {
        return b -> b == this.getBlockState().getBlock();
    }

    protected Map<BlockPos, Predicate<BlockState>> getRotatedPattern() {
        if (getBlockState().getValues().containsKey(RotatableMachineBlock.FACING)) {
            switch (getBlockState().get(RotatableMachineBlock.FACING)) {
                case SOUTH:
                    return getPattern().getPatternMap().entrySet().stream().map(e -> new Pair<>(new BlockPos(e.getKey().getX() * -1, e.getKey().getY(), e.getKey().getZ() * -1), e.getValue())).collect(Collectors.toMap(Pair::getKey, Pair::getValue));
                case EAST:
                    return getPattern().getPatternMap().entrySet().stream().map(e -> new Pair<>(new BlockPos(e.getKey().getZ() * -1, e.getKey().getY(), e.getKey().getX()), e.getValue())).collect(Collectors.toMap(Pair::getKey, Pair::getValue));
                case WEST:
                    return getPattern().getPatternMap().entrySet().stream().map(e -> new Pair<>(new BlockPos(e.getKey().getZ(), e.getKey().getY(), e.getKey().getX() * -1), e.getValue())).collect(Collectors.toMap(Pair::getKey, Pair::getValue));
            }
        }
        return getPattern().getPatternMap();
    }
}

File: src/main/java/com/emosewapixel/pixellib/multiblocks/MultiblockTE.kt
package com.emosewapixel.pixellib.multiblocks

import com.emosewapixel.pixellib.blocks.RotatableMachineBlock
import com.emosewapixel.pixellib.recipes.AbstractRecipeList
import com.emosewapixel.pixellib.recipes.SimpleMachineRecipe
import com.emosewapixel.pixellib.tiles.AbstractRecipeBasedTE
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.tileentity.TileEntityType
import net.minecraft.util.Direction
import net.minecraft.util.math.BlockPos
import java.util.function.Predicate

//Multiblock TEs are TEs that require a multiblock structure to work
abstract class MultiblockTE<T : SimpleMachineRecipe>(type: TileEntityType<*>, recipeList: AbstractRecipeList<T, *>) : AbstractRecipeBasedTE<T>(type, recipeList) {
    protected lateinit var partList: List<IMultiblockPart>

    protected abstract val pattern: MultiblockPatternBuilder

    protected val isValidStructure: Boolean
        get() {
            val pattern = rotatedPattern
            val posInPattern = pattern.entries.filter { it.value.test(blockState) }.map { it.key }.first()

            partList = pattern.keys.filter { world!!.getTileEntity(it) is IMultiblockPart }.map { world!!.getTileEntity(it) as IMultiblockPart }

            return pattern.all { (key, value) -> value.test(world!!.getBlockState(getPos().subtract(posInPattern).add(key))) }
        }

    protected val rotatedPattern: Map<BlockPos, Predicate<BlockState>>
        get() {
            if (blockState.values.containsKey(RotatableMachineBlock.FACING)) {
                when (blockState.get(RotatableMachineBlock.FACING)) {
                    Direction.SOUTH -> return pattern.patternMap.entries.map { e -> Pair(BlockPos(e.key.x * -1, e.key.y, e.key.z * -1), e.value) }.toMap()
                    Direction.EAST -> return pattern.patternMap.entries.map { e -> Pair(BlockPos(e.key.z * -1, e.key.y, e.key.x), e.value) }.toMap()
                    Direction.WEST -> return pattern.patternMap.entries.map { e -> Pair(BlockPos(e.key.z, e.key.y, e.key.x * -1), e.value) }.toMap()
                }
            }
            return pattern.patternMap
        }

    protected fun selfPredicate() = Predicate<Block> { it === this.blockState.block }
}